solidity/contracts/modules/MultisigModule.sol

```diff
@@ -0,0 +1,349 @@
+     * a checkpoint, and verifies the merkle proof of `_message` against that
+     * checkpoint.
+     * @param _metadata ABI encoded module metadata (see MultisigModuleMetadata.sol)
+     * @param _message Formatted Hyperlane message (see Message.sol).
+     */
+    function verify(bytes calldata _metadata, bytes calldata _message)
+        public
+        view
+        returns (bool)
+    {
+        require(_verifyMerkleProof(_metadata, _message), "!merkle");
+        require(_verifyValidatorSignatures(_metadata, _message), "!sigs");
+        return true;
+    }
+
+    /**
+     * @notice Gets the current validator set, sorted by ascending address.
+     * @param _domain The remote domain of the validator set.
+     * @return The addresses of the validator set.
+     */
+    function validators(uint32 _domain) public view returns (address[] memory) {
+        EnumerableSet.AddressSet storage _validatorSet = validatorSet[_domain];
+        uint256 _validatorCount = _validatorSet.length();
+        address[] memory _validators = new address[](_validatorCount);
+        address _prev = address(0);
+        for (uint256 i = 0; i < _validatorCount; i++) {
+            // Max address
+            address _next = address(0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF);
+            for (uint256 j = 0; j < _validatorCount; j++) {
+                address _validator = _validatorSet.at(j);
+                if (_prev < _validator && _validator < _next) {
+                    _next = _validator;
+                }
+            }
+            _validators[i] = _next;
+            _prev = _next;
+        }
+        return _validators;
+    }
+
+    /**
+     * @notice Returns the number of validators enrolled in the validator set.
+     * @param _domain The remote domain of the validator set.
+     * @return The number of validators enrolled in the validator set.
+     */
+    function validatorCount(uint32 _domain) public view returns (uint256) {
+        return validatorSet[_domain].length();
+    }
+
+    // ============ Internal Functions ============
+
+    /**
+     * @notice Updates the commitment to the validator set for `_domain`.
+     * @param _domain The remote domain of the validator set.
+     * @return The commitment to the validator set for `_domain`.
+     */
+    function _updateCommitment(uint32 _domain) internal returns (bytes32) {
+        address[] memory _validators = validators(_domain);
+        uint256 _threshold = threshold[_domain];
+        bytes32 _commitment = keccak256(
+            abi.encodePacked(_threshold, _validators)
+        );
+        commitment[_domain] = _commitment;
+        return _commitment;
+    }
+
+    /**
+     * @notice Verifies the merkle proof of `_message` against the provided
+     * checkpoint.
+     * @param _metadata ABI encoded module metadata (see MultisigModuleMetadata.sol)
+     * @param _message Formatted Hyperlane message (see Message.sol).
+     */
+    function _verifyMerkleProof(
```

Comment:
This is most likely premature, but this could be nice in a base class for ISMs maybe? IIUC an optimistic ISM could then subclass from both?

Reply:
Agreed this is premature, let's move it if/when when it needs to be shared